### PR TITLE
Upgrade IPC

### PR DESCRIPTION
**Pull request by junha1**
This PR contains multiple commits. 
Most of them are for fixing bugs, and others are preparations for FML PR.

This change is due to MacOS slightly different behavior of using domain socket. 
I must use `connect()` to use `shutdown()` in MacOS. There are some minor differences but this is the main background of this PR.

### CODE REVIEW

File: basesandbox/src/ipc/multiplex.rs [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+impl Multiplexer {
+    pub fn create<F: Forward, S: IpcSend + 'static, R: IpcRecv + 'static>(
+        ipc_send: S,
+        ipc_recv: R,
+        multiplex: usize,
+        channel_capacity: usize,
+    ) -> (Vec<(Sender, Receiver)>, Self) {
+        let (sender_send, sender_recv) = bounded(channel_capacity);
+        let mut receiver_sends = Vec::<Sender>::new();
+        let mut channel_ends = Vec::<(Sender, Receiver)>::new();
```
Comment (majecty) [inline basesandbox/src/ipc/multiplex.rs:66]: How about introduce a ChannelEnd struct?
It is hard to read what is the role of Sender and Receiver.

File: basesandbox/src/execution/executor.rs [hunk 2]
```diff
@@ -66,6 +66,8 @@ fn get_function_pool(key: &str) -> ThreadAsProcesss {
 }
 
 pub fn add_function_pool(key: String, f: ThreadAsProcesss) {
+    // this is ok to return Some(_).
+    // we don't really have to have a bothersome synchronization to assure the unique registeration of this
```
Comment (majecty) [inline basesandbox/src/execution/executor.rs:70]:
```suggestion
    // This is ok to return Some(_).
    // We don't really have to have a bothersome synchronization to assure the unique registeration of this
```

Comment (majecty) [reply basesandbox/src/execution/executor.rs:70]: I didn't understand this comment. Could you explain it to me?

File: basesandbox/src/ipc/domain_socket.rs [hunk 3]
```diff
@@ -116,14 +108,24 @@ impl Ipc for DomainSocket {
...
+        let socket = UnixDatagram::bind(&address_src).unwrap();
+        (|| {
+            for _ in 0..100 {
+                if socket.connect(&address_dst).is_ok() {
+                    return Ok(())
+                }
+                std::thread::sleep(std::time::Duration::from_millis(10));
+            }
+            Err("Failed to establish domain socket")
+        })()
```
[the remark this answers is not in the document]
Comment (junha1) [reply basesandbox/src/ipc/domain_socket.rs:119]: Why is using additional `mut success` better than lambda?

File: basesandbox/src/ipc/multiplex.rs [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+use std::thread;
+
+type Sender = crossbeam::channel::Sender<Vec<u8>>;
+type Receiver = crossbeam::channel::Receiver<Vec<u8>>;
+
+pub trait Forward {
+    fn forward(data: &[u8]) -> usize;
+}
+
+fn sender<S: IpcSend>(send: S, recv: Receiver) {
```
[the remark this answers is not in the document]
Comment (junha1) [reply basesandbox/src/ipc/multiplex.rs:28]: Any suggestion?

File: basesandbox/src/ipc/multiplex.rs (hunk 1 above)
Comment (junha1) [reply basesandbox/src/ipc/multiplex.rs:66]: Well even I if I do so the field name will be just `sender` and `receiver`

File: basesandbox/src/ipc/domain_socket.rs [hunk 5]
```diff
@@ -98,16 +98,16 @@ impl Ipc for DomainSocket {
...
+            serde_cbor::to_vec(&(false, &address_client, &address_server)).unwrap(),
         )
     }
 
     type SendHalf = DomainSocketSend;
     type RecvHalf = DomainSocketRecv;
 
     fn new(data: Vec<u8>) -> Self {
-        let (address_src, address_dst): (String, String) = serde_cbor::from_slice(&data).unwrap();
+        let (am_i_server, address_src, address_dst): (bool, String, String) = serde_cbor::from_slice(&data).unwrap();
```
[the remark this answers is not in the document]
Comment (junha1) [reply basesandbox/src/ipc/domain_socket.rs:109]: Note that that struct must be serde `Serialize` and `Deserialize`. And this tuple is only used in this module, only two places. Is it really worth it to refactor out with struct now?

File: basesandbox/src/ipc/multiplex.rs [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+}
+
+fn receiver<F: Forward, R: IpcRecv>(recv: R, send: Vec<Sender>) {
+    loop {
+        let data = match recv.recv(None) {
+            Err(RecvError::TimeOut) => panic!(),
+            Err(RecvError::Termination) => return,
+            Ok(x) => x,
+        };
+        send[F::forward(&data)].send(data).unwrap();
```
Comment (majecty) [inline basesandbox/src/ipc/multiplex.rs:46]: I understand the Multiplexer.
IMO, using number index to route packets is not intuitive.

How about this interface? What do you think about it?

```rust
let server_recv;
let client_recv;
Multiplexer::create(ipc_send, ipc_recv, move |data| {
  let header = PacketHeader::new(&data);
  if header.slot >= SLOT_CALL_OR_RETURN_INDICATOR {
    server_recv.send(data);
  } else {
    client_recv.send(data);
  }
}
```

Comment (junha1) [reply basesandbox/src/ipc/multiplex.rs:46]: I don't get it. Where do you want me to put this code?

File: basesandbox/src/ipc/domain_socket.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (majecty) [reply basesandbox/src/ipc/domain_socket.rs:119]: Yes. Creating closure and running it is not a natural way to use the lambda syntax.
When you write a hacky code, there should be a huge benefit for it.
IMO, using `mut success` is more natural way, and using closure does not have much benefit than `mut success`.

[the remark this answers is not in the document]
Comment (junha1) [reply basesandbox/src/ipc/domain_socket.rs:119]: Well using less mutable variables is a good thing.
However if you care the readability, I'll apply your suggestion.

[the remark this answers is not in the document]
Comment (majecty) [reply basesandbox/src/ipc/domain_socket.rs:119]: Yes, using mutable variables less is a good practice in general. Mutable variables make code hard to run in parallel and make code complex if the variable could have many states.
However, in this case, we don't need parallelism and mutable does not make code complex. In this case, using a mutable variable make code better.

[the remark this answers is not in the document]
Comment (junha1) [reply basesandbox/src/ipc/domain_socket.rs:119]: ok

File: basesandbox/src/ipc/intra.rs [hunk 7]
```diff
@@ -111,11 +117,21 @@ impl Ipc for Intra {
...
+            send: send_server.clone(),
+            recv: recv_server,
+            send_for_termination: send_client.clone(),
+        });
+        add_ends(key_client.clone(), RegisteredIpcEnds {
+            is_server: false,
+            send: send_client,
+            recv: recv_client,
+            send_for_termination: send_server,
+        });
```
Comment (majecty) [inline basesandbox/src/ipc/intra.rs:134]: Thanks. It is easier to understand.

File: basesandbox/src/ipc/multiplex.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (majecty) [reply basesandbox/src/ipc/multiplex.rs:28]: How about `run_sender_loop`?

File: basesandbox/src/ipc/multiplex.rs [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+        let data = recv.recv().unwrap();
+        // special exit flag
+        if data.is_empty() {
+            break
+        }
+        send.send(&data);
+    }
+}
+
+fn receiver<F: Forward, R: IpcRecv>(recv: R, send: Vec<Sender>) {
```
[the remark this answers is not in the document]
Comment (majecty) [reply basesandbox/src/ipc/multiplex.rs:39]: How about `run_receiver_loop`?

File: basesandbox/src/ipc/multiplex.rs (hunk 6 above)
Comment (majecty) [reply basesandbox/src/ipc/multiplex.rs:46]: We talked about it in a call. We concluded that add a unit test to show how to use the Multiplexer for the reader.

File: basesandbox/src/ipc/multiplex.rs (hunk 1 above)
Comment (majecty) [reply basesandbox/src/ipc/multiplex.rs:66]: OK

File: basesandbox/src/ipc/domain_socket.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (majecty) [reply basesandbox/src/ipc/domain_socket.rs:109]: Let's do it when there are more fields.

File: basesandbox/src/ipc/multiplex.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (dynaxis) [reply basesandbox/src/ipc/multiplex.rs:28]: Actually there are cases where noun names like `senderLoop` or `mainLoop` makes sense, IMHO.
But `sender` is somewhat insufficient to give the right sense of what it does.

If we want to stick to a verb name for a function, then `run_sender_loop` or `enter_sender_loop` seems good to me.

[the remark this answers is not in the document]
Comment (majecty) [reply basesandbox/src/ipc/multiplex.rs:28]: `sender_loop` can be a good option.

File: basesandbox/tests/intergration_test.rs [hunk 9]
```diff
@@ -140,3 +138,91 @@ fn execute_simple_intra_massive() {
...
+    s1_2.send(b"04".to_vec()).unwrap();
+
+    assert_eq!(r2_1.recv().unwrap(), b"11");
+    assert_eq!(r1_2.recv().unwrap(), b"03");
+    assert_eq!(r2_2.recv().unwrap(), b"04");
+    assert_eq!(r1_1.recv().unwrap(), b"12");
+
+    // we have to drop multiplexer itself first
+    drop(multiplxer1);
+    drop(multiplxer2);
```
Comment (majecty) [inline basesandbox/tests/intergration_test.rs:227]: How about changing indexes a bit?

```suggestion
    let (mut multiplexed1, multiplxer1) = Multiplexer::create::<TestForward, _, _>(s1, r1, 2, 100);
    let (mut multiplexed2, multiplxer2) = Multiplexer::create::<TestForward, _, _>(s2, r2, 2, 100);

    // multiplxed channel 1
    let (s1_1, r1_1) = multiplexed1.pop().unwrap();
    let (s2_1, r2_1) = multiplexed2.pop().unwrap();

    // multiplxed channel 0
    let (s1_0, r1_0) = multiplexed1.pop().unwrap();
    let (s2_0, r2_0) = multiplexed2.pop().unwrap();

    s1_1.send(b"11".to_vec()).unwrap();
    assert_eq!(r2_1.recv().unwrap(), b"11");

    s2_1.send(b"12".to_vec()).unwrap();
    assert_eq!(r1_1.recv().unwrap(), b"12");
    
    s2_0.send(b"03".to_vec()).unwrap();
    assert_eq!(r1_0.recv().unwrap(), b"03");
    
    s1_0.send(b"04".to_vec()).unwrap();
    assert_eq!(r2_0.recv().unwrap(), b"04");
    
    // we have to drop multiplexer itself first
    drop(multiplxer1);
    drop(multiplxer2);
```

File: basesandbox/src/ipc.rs [hunk 10]
```diff
@@ -17,6 +17,7 @@
 pub mod domain_socket;
 pub mod intra;
 pub mod multiplex;
+pub mod servo;
```
[the remark this answers is not in the document]
Comment (junha1) [reply basesandbox/src/ipc.rs:20]: @majecty I'll use `servo_ipc` since `ipc_channel` is even more confusing

**Pull request merged**